### PR TITLE
Add has_path and a minimal sentence generator.

### PR DESCRIPTION
**Pull request by ltratt**
It's best to review the three commits separately rather than the overall PR: `has_path` is unrelated to the other two functions. [At one point I thought I needed `has_path`, but currently I don't. I'm not sure whether I should keep it around or not. Any thoughts?]

`min_sentence` returns a single minimal sentence match a non-terminal; `min_sentences` returns all minimal sentences matching a non-terminal.

### CODE REVIEW

File: src/lib/yacc/grammar.rs [hunk 1]
```diff
@@ -363,6 +363,43 @@ impl YaccGrammar {
...
+                            }
+                            if !seen[usize::from(nt_idx)] {
+                                todo[usize::from(nt_idx)] = true;
+                            }
+                        }
+                    }
+                }
+            }
+            if empty {
+                break;
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs]: Why don't we immediately `return false` here instead of leaving the loop and then returning false?

Comment (ltratt) [reply src/lib/yacc/grammar.rs]: Good point!

**Comment by ptersilie** [comment]
Second commit: Did you mean "special use case" instead of "specialist use case"?

File: src/lib/yacc/grammar.rs [hunk 2]
```diff
@@ -400,6 +400,17 @@ impl YaccGrammar {
         }
         false
     }
+
+    /// Return a `SentenceGenerator` which can then generate minimal sentences for any non-term
+    /// based on the user-defined `term_cost` function which gives the associated cost for
+    /// generating each terminal (where the cost must be greater than 0). Note that multiple
+    /// terminals can have the same score. The simplest cost function is thus `|_| 1`.
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs:406]: What does `|_|` mean?

File: src/lib/yacc/grammar.rs [hunk 3]
```diff
@@ -421,6 +432,171 @@ impl Grammar for YaccGrammar {
...
+impl<'a> SentenceGenerator<'a> {
+    fn new<F>(grm: &YaccGrammar, term_cost: F) -> SentenceGenerator
+        where F: Fn(TIdx) -> u64
+    {
+        let nt_costs = nonterm_costs(grm, &term_cost);
+        let mut term_costs = Vec::with_capacity(grm.terms_len());
+        for i in 0..grm.terms_len() {
+            term_costs.push(term_cost(TIdx::from(i)));
+        }
+        SentenceGenerator{grm, term_costs, nonterm_costs: nt_costs}
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs:469]: Why are you not initialising this with `SentenceGenerator{grm, nt_costs, term_costs}` (why are you swapping `nt_costs` and `term_costs` around) ?

File: src/lib/yacc/grammar.rs [hunk 4]
```diff
@@ -421,6 +432,171 @@ impl Grammar for YaccGrammar {
...
+    }
+}
+
+/// Return the cost of a minimal string for each nonterminal in this grammar. The cost of a
+/// terminal is specified by the user-defined `term_cost` function.
+fn nonterm_costs<F>(grm: &YaccGrammar, term_cost: F) -> Vec<u64>
+               where F: Fn(TIdx) -> u64
+{
+    // We use a simple(ish) fixed-point algorithm to determine costs. We maintain two lists
+    // "costs" and "done". An integer costs[i] starts at 0 monotonically increments
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs]: **and** monotonically increments

File: src/lib/yacc/grammar.rs [hunk 5]
```diff
@@ -421,6 +432,171 @@ impl Grammar for YaccGrammar {
...
+    // until done[i] is true, at which point costs[i] value is fixed. We also use the done
+    // list as a simple "todo" list: whilst there is at least one false value in done, there is
+    // still work to do.
+    //
+    // On each iteration of the loop, we examine each non-terminal in the todo list to see if
+    // we can get a better idea of its true cost. Some are trivial:
+    //   * A non-terminal with an empty production immediately has a cost of 0.
+    //   * Non-terminals whose productions which don't reference any non-terminals
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs]: delete **which**

File: src/lib/yacc/grammar.rs (hunk 2 above)
Comment (ltratt) [reply src/lib/yacc/grammar.rs:406]: It's a closure which takes one argument, but in this case the closure doesn't reference it, so `_` means "throw the argument away".

File: src/lib/yacc/grammar.rs (hunk 3 above)
Comment (ltratt) [reply src/lib/yacc/grammar.rs:469]: It's the way Rust struct initialisation works (in recent versions of rustc): if a field and a variable name *happen* to share the name, you don't need to specify both; but in this case, they don't.

File: src/lib/yacc/grammar.rs (hunk 4 above)
Comment (ltratt) [reply src/lib/yacc/grammar.rs]: Eek.

**Comment by ltratt** [comment]
I think I've addressed those comments, although I'm guessing you've yet to look at the third commit?

File: src/lib/yacc/grammar.rs [hunk 6]
```diff
@@ -684,4 +860,69 @@ mod test {
         assert!(!grm.has_path(b_nt_idx, a_nt_idx));
         assert!(!grm.has_path(c_nt_idx, a_nt_idx));
     }
+
+    #[test]
+    fn test_nonterm_costs() {
+        let grm = yacc_grm(YaccKind::Original, &"
```
Comment (ptersilie) [inline src/lib/yacc/grammar.rs]: This test could be slightly improved by having a rule `E: 'x' A | 'x' 'y'` to check that the algorithm tests `'x' A` as well which is shorter than `xy`.

File: src/lib/yacc/grammar.rs (hunk 2 above)
Comment (ptersilie) [reply src/lib/yacc/grammar.rs:406]: I see. I first thought you are trying yourself in ascii art. `¯\_(ツ)_/¯`

File: src/lib/yacc/grammar.rs (hunk 3 above)
Comment (ptersilie) [reply src/lib/yacc/grammar.rs:469]: Ah so the arguments are unordered? Meaning `SentenceGenerator{grm, nt_costs, term_costs}` won't work?

File: src/lib/yacc/grammar.rs (hunk 2 above)
Comment (ltratt) [reply src/lib/yacc/grammar.rs:406]: If so, I would have represented my brain as `.` (to scale).

File: src/lib/yacc/grammar.rs (hunk 3 above)
Comment (ltratt) [reply src/lib/yacc/grammar.rs:469]: Exactly.

**Comment by ptersilie** [comment]
Just a few more comments. Rest looks fine.

**Comment by ltratt** [comment]
OK, I *think* the new commits cover all your comments. Please let me know if they don't.

**Review by ptersilie — CHANGES_REQUESTED** [review]
Looking good now. Will merge after squashing!

**Comment by ptersilie** [comment]
Oh, it seems review messages are not forwarded to Mattermost. I assume that's why you haven't seen this yet. I will have to update our plugin then.

**Comment by ltratt** [comment]
Squashed.

**Review by ptersilie — APPROVED** [review]

**Pull request merged**